File: CS/Form1.cs
using DevExpress.XtraEditors;
using DevExpress.XtraEditors.Repository;
using DevExpress.XtraGrid;
using DevExpress.XtraGrid.Views.Grid;
using DevExpress.XtraGrid.Views.Grid.ViewInfo;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;

namespace TestGridLookUpEditCSV {
    public partial class Form1 : DevExpress.XtraEditors.XtraForm {
        public Form1() {
            InitializeComponent();
            InitGrid();
            RepositoryItemPopupContainerEdit ri = new RepositoryItemPopupContainerEdit();
            ri.QueryResultValue += ri_QueryResultValue;
            ri.EditValueChanged += ri_EditValueChanged;
            ri.Popup += ri_Popup;
            ri.TextEditStyle = DevExpress.XtraEditors.Controls.TextEditStyles.Standard;
            PopupContainerControl popupControl = new PopupContainerControl();
            popupControl.Controls.Add(popupGridControl1);
            popupControl.Size = new System.Drawing.Size(500, 300);
            ri.PopupControl = popupControl;
            gridView1.Columns["Name"].ColumnEdit = ri;

            popupContainerEdit1.Properties.PopupControl = popupContainerControl1;
            popupContainerEdit1.Properties.QueryResultValue += Properties_QueryResultValue;
            popupContainerEdit1.Properties.EditValueChanged += Properties_EditValueChanged;
            popupContainerEdit1.Properties.Popup += Properties_Popup;

        }

        void Properties_Popup(object sender, EventArgs e) {
            PopupContainerEdit edit = sender as PopupContainerEdit;
            if(!edit.IsPopupOpen)
                edit.ShowPopup();
            UpdateSelection(edit, popupGridView2);
        }

        void Properties_EditValueChanged(object sender, EventArgs e) {
            PopupContainerEdit edit = sender as PopupContainerEdit;
            UpdateSelection(edit, popupGridView2);
        }

        void Properties_QueryResultValue(object sender, DevExpress.XtraEditors.Controls.QueryResultValueEventArgs e) {
            int[] selectedRows = popupGridView2.GetSelectedRows();
            List<string> values = new List<string>();
            foreach(int rowHandle in selectedRows) {
                values.Add(popupGridView2.GetRowCellValue(rowHandle, "Name").ToString());
            }
            string csv = String.Join(", ", values);
            e.Value = csv;
        }

        void ri_Popup(object sender, EventArgs e) {
            PopupContainerEdit edit = sender as PopupContainerEdit;
            UpdateSelection(edit, popupGridView1);
        }

        void ri_EditValueChanged(object sender, EventArgs e) {
            PopupContainerEdit edit = sender as PopupContainerEdit;
            if(!edit.IsPopupOpen)
                edit.ShowPopup();
            UpdateSelection(edit, popupGridView1);
        }

        void ri_QueryResultValue(object sender, DevExpress.XtraEditors.Controls.QueryResultValueEventArgs e) {
            int[] selectedRows = popupGridView1.GetSelectedRows();
            List<string> values = new List<string>();
            foreach(int rowHandle in selectedRows) {
                values.Add(popupGridView1.GetRowCellValue(rowHandle, "Name").ToString());
            }
            string csv = String.Join(", ", values);
            e.Value = csv;
        }

        private void UpdateSelection(PopupContainerEdit edit, GridView view) {
            view.BeginSelection();
            view.ClearSelection();
            if(edit != null)
                if(edit.EditValue != null) {
                    edit.Focus();
                    List<int> selection = GetSelection(edit.EditValue.ToString().Split(new string[] { ", " }, StringSplitOptions.None), "Name", view);
                    foreach(int rowHandle in selection) {
                        view.SelectRow(rowHandle);
                    }
                }
            view.EndSelection();
        }

        private List<int> GetSelection(string[] values, string fieldName, GridView view) {
            List<int> selection = new List<int>();
            foreach(string val in values) {
                for(int i = 0; i < view.RowCount; i++) {
                    if(view.GetRowCellValue(i, fieldName).ToString() == val)
                        selection.Add(i);
                }
            }
            return selection;
        }



        private void InitGrid() {
            popupGridControl1.DataSource = CreateList(5);
            gridControl2.DataSource = CreateList(5);
            popupGridControl2.DataSource = CreateList(5);
        }

        BindingList<Customer> CreateList(int count) {
            BindingList<Customer> list = new BindingList<Customer>();
            for(int i = 0; i < count; i++) {
                list.Add(new Customer(i, "Name" + i, "Info" + i));
            }
            return list;
        }

    }
}

File: VB/Form1.vb
Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.Repository
Imports DevExpress.XtraGrid.Views.Grid
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Drawing

Namespace TestGridLookUpEditCSV

    Public Partial Class Form1
        Inherits XtraForm

        Public Sub New()
            InitializeComponent()
            InitGrid()
            Dim ri As RepositoryItemPopupContainerEdit = New RepositoryItemPopupContainerEdit()
            AddHandler ri.QueryResultValue, AddressOf ri_QueryResultValue
            AddHandler ri.EditValueChanged, AddressOf ri_EditValueChanged
            AddHandler ri.Popup, AddressOf ri_Popup
            ri.TextEditStyle = DevExpress.XtraEditors.Controls.TextEditStyles.Standard
            Dim popupControl As PopupContainerControl = New PopupContainerControl()
            popupControl.Controls.Add(popupGridControl1)
            popupControl.Size = New System.Drawing.Size(500, 300)
            ri.PopupControl = popupControl
            gridControl2.RepositoryItems.Add(ri)
            gridView1.Columns("Name").ColumnEdit = ri
            popupContainerEdit1.Properties.PopupControl = popupContainerControl1
            AddHandler popupContainerEdit1.Properties.QueryResultValue, AddressOf Properties_QueryResultValue
            AddHandler popupContainerEdit1.Properties.EditValueChanged, AddressOf Properties_EditValueChanged
            AddHandler popupContainerEdit1.Properties.Popup, AddressOf Properties_Popup
        End Sub

        Private Sub Properties_Popup(ByVal sender As Object, ByVal e As EventArgs)
            Dim edit As PopupContainerEdit = TryCast(sender, PopupContainerEdit)
            If Not edit.IsPopupOpen Then edit.ShowPopup()
            UpdateSelection(edit, popupGridView2)
        End Sub

        Private Sub Properties_EditValueChanged(ByVal sender As Object, ByVal e As EventArgs)
            Dim edit As PopupContainerEdit = TryCast(sender, PopupContainerEdit)
            UpdateSelection(edit, popupGridView2)
        End Sub

        Private Sub Properties_QueryResultValue(ByVal sender As Object, ByVal e As DevExpress.XtraEditors.Controls.QueryResultValueEventArgs)
            Dim selectedRows As Integer() = popupGridView2.GetSelectedRows()
            Dim values As List(Of String) = New List(Of String)()
            For Each rowHandle As Integer In selectedRows
                values.Add(popupGridView2.GetRowCellValue(rowHandle, "Name").ToString())
            Next

            Dim csv As String = String.Join(", ", values)
            e.Value = csv
        End Sub

        Private Sub ri_Popup(ByVal sender As Object, ByVal e As EventArgs)
            Dim edit As PopupContainerEdit = TryCast(sender, PopupContainerEdit)
            UpdateSelection(edit, popupGridView1)
        End Sub

        Private Sub ri_EditValueChanged(ByVal sender As Object, ByVal e As EventArgs)
            Dim edit As PopupContainerEdit = TryCast(sender, PopupContainerEdit)
            If Not edit.IsPopupOpen Then edit.ShowPopup()
            UpdateSelection(edit, popupGridView1)
        End Sub

        Private Sub ri_QueryResultValue(ByVal sender As Object, ByVal e As DevExpress.XtraEditors.Controls.QueryResultValueEventArgs)
            Dim selectedRows As Integer() = popupGridView1.GetSelectedRows()
            Dim values As List(Of String) = New List(Of String)()
            For Each rowHandle As Integer In selectedRows
                values.Add(popupGridView1.GetRowCellValue(rowHandle, "Name").ToString())
            Next

            Dim csv As String = String.Join(", ", values)
            e.Value = csv
        End Sub

        Private Sub UpdateSelection(ByVal edit As PopupContainerEdit, ByVal view As GridView)
            view.BeginSelection()
            view.ClearSelection()
            If edit IsNot Nothing Then
                If edit.EditValue IsNot Nothing Then
                    edit.Focus()
                    Dim selection As List(Of Integer) = GetSelection(edit.EditValue.ToString().Split(New String() {", "}, StringSplitOptions.None), "Name", view)
                    For Each rowHandle As Integer In selection
                        view.SelectRow(rowHandle)
                    Next
                End If
            End If

            view.EndSelection()
        End Sub

        Private Function GetSelection(ByVal values As String(), ByVal fieldName As String, ByVal view As GridView) As List(Of Integer)
            Dim selection As List(Of Integer) = New List(Of Integer)()
            For Each val As String In values
                For i As Integer = 0 To view.RowCount - 1
                    If Equals(view.GetRowCellValue(i, fieldName).ToString(), val) Then selection.Add(i)
                Next
            Next

            Return selection
        End Function

        Private Sub InitGrid()
            popupGridControl1.DataSource = CreateList(5)
            gridControl2.DataSource = CreateList(5)
            popupGridControl2.DataSource = CreateList(5)
        End Sub

        Private Function CreateList(ByVal count As Integer) As BindingList(Of Customer)
            Dim list As BindingList(Of Customer) = New BindingList(Of Customer)()
            For i As Integer = 0 To count - 1
                list.Add(New Customer(i, "Name" & i, "Info" & i))
            Next

            Return list
        End Function
    End Class
End Namespace
